### ISSUE TITLE
Love the project!! Some ideas/questions:

### ISSUE THREAD

**Issue by ezuk** [sboesebeck/ErgodoxLayoutGenerator#2]
Hey @sboesebeck :)

Very cool stuff. Moving the discussion here as you suggested.

1) ASCII legends: Currently, keymaps produced in qmk (even with the little Python script, are you familiar with it?) have an ASCII legend of each layer built into the keymap as comments. IMHO it would be cool to preserve this with ELG, because ELG can already read keymaps, so it would make it totally interoperable with the existing keymaps. What do you think?

2) Once the project is ready for it, I would love to make a video for you, showing its use and installation.  Can publish it on my YouTube channel and share with the ErgoDox EZ community in general.

3) I'm dreaming here, but it would be mega-cool if we had one setup bundle that basically installed ELG + the build env. So you next-next-next your way through something, then tada, you have a graphical UI to set up your ErgoDox EZ. Maybe the simplest thing here would be a downloadable VirtualBox with ELG + build env running. Any thoughts on this?

**Comment by sboesebeck** [comment]
Hey Erez....

1) yes, I was thinking about this as well... I postponed it for now, was not on the top prio. But I'll think of something.
2) sure, that would be awesome!
3) I thought about that as well. Thing is, due to the different OS this gets complicated. Did not have the Idea yet to package it into a virtual machine... Sounds like a very cool idea!

Lets elaborate on 3 a bit more...

The goal would be to get a GUI that actually is able to do everything. the whole process, reading, writing, compiling and flashing. If we want to achieve that goal, we actually need to add the Teensy tool (or better the commandline version) to the package as well. Then you have an all in one tool for the whole thing!

And the more thoughts i give this, I think the only way to get that in a simple to install package is by some virualization. Maybe a VirtualBox is really the best solution. I really love this idea...

**Comment by ezuk** [comment]
I'm glad you like it! I'm thinking of some stripped-down version of Ubuntu
with a very basic window manager and one icon (to run the tool).

**Comment by sboesebeck** [comment]
I did something similar to that already a couple of years ago. This acutally did not use a windowmanager at all - if you only have one window- why use a window manager?

has it's drawbacks though...

I would work on the teensy integration first. Then let's see...

cheers,

Stephan

**Comment by ezuk** [comment]
Sounds good!

**Comment by sboesebeck** [comment]
Just created the first final release 1.0

this does create an ascii legend to the keymap.c file.

I was also thinking about the thing with the virtual box. Unfortunately, this would blow up things _a lot_. my Image file is now about 5GB in size. This is one hell of a download. Maybe someone could help in creating it. The most complex thing is acutally getting X-Windows to run properly. But still, the download will be around 3GB anyway...

Also: I would stick to the teensy loader app, the CLI is a bit too error prone and I'd have to re-implement about the same thing as in the teensy loader app, only in java. No benefit from it.

I'd suggest the following: I close this issue. We try to get more experiences with 1.0 and see how things are going. I will upload the VM somewhere, so somebody else can have a closer look at it too...

**Comment by ezuk** [comment]
I am totally fine with grabbing a 5GB VM (VirtualBox I assume?). Send me the link and I'll download and test it out, will come back with feedback. :)

**Comment by sboesebeck** [comment]
Yes, it's VirtualBox... I'll upload it somewhere later today....

**Comment by sboesebeck** [comment]
Just uploaded a virtualBox image to my server... everything is installed, that is needed to compile and flash your firmware.
- based on ubuntu server makes it smaller
- login is "elg" password "elg"
- when logging in, the current sources of the qmk-firmware and the ErgodoxLayoutGenerator are downloaded / updated (git pull)
- the ELG is compiled (from master branch) - this actually contains a fix for the 1.0 which was unfortunately unable to compile on linux.
- teensy app is started
- ELG is started...

so, everything should work then. I got some issues though... there seem to be some issues with XDM when direktly starting it from boot. I could not start x from console for whatever reason. Switching to VC1 (text mode) and back to VC7 usually helps...

you can download it at https://www.boesebeck.name/elg-srv.zip https://www.boesebeck.name/elg-srv.zip  1.5GB zipped, ~3.5GB unzipped

hope you like it...

Cheers,

Stephan

**Comment by ezuk** [comment]
Downloading it now, will try it out in the next few days and report!

**Comment by ezuk** [comment]
Okay, I'm back with a little bit of feedback!

1) All in all, this is super cool. The boot process is nice, the way it pulls the repo.

2) The boot issue you described happened to me too, and would be a dealbreaker for most, I think.

3) The app itself is lovely!

4) Compilation failed when I tried it, but I couldn't copy the error trace. I recommend installing VirtualBox Guest additions. Maybe we don't need the super minimal window manager, can just do with Unity? Or do you not like Ubuntu? 

5) Would be good to do away with any login requirement, just have the thing log in and start automatically.

What do you think?

**Comment by sboesebeck** [comment]
Hi Erez,

Great that you like it!

I solved the boot issue and the auto login by switching to a different 
display manager. Now, that boots right up into x, pulls all current 
sources and compiles them.

The VirtualBoxGuest additions are installed btw. They make it possible 
to resize the vbox-window and have the screen resolution on X11 change 
accordingly - does that work on your side (it takes some time to adjust 
though, about 2-3 seconds)? the copy'n'paste thing did not work, as it 
was disabled in the settings for this VM -it works now.

The compilation worked here, I don't know what went wrong on your side. 
Is that only failing in Virtual Box?

I did not want to have a full blown Ubuntu installation for running one 
single app, I tried to keep it small - I installed everything on a 
Desktop Ubuntu, which then was 5GB in size. I thought, that is way to 
much space, which is not really needed.

I uploaded a new version of the VM, which does boot right into X, starts 
everything automatically, and I added a link to the documentation of the 
ELG in the IceWM-Menu...

**Comment by ezuk** [comment]
Great! Grabbing the new version now and will report.

**Comment by sboesebeck** [comment]
Just to mention: there is no support for the FN-functionality yet. As I see it, this is doing the same as macros. Do you think, this is necessary?

It's a bit of a problem, that you can do the same things in so many different ways in the firmware. The ELG will always use some "minimum" approach - or at least an 80% kind of thing. 

Right now, I have the problem with the "zweihander_osx"-keymap, which uses FN1 to do a momentary switch to a layer. This could also be achieved by using `MO()`, which is supported by the ELG. As I see it- or is there a difference in the two?

**Comment by ezuk** [comment]
> The ELG will always use some "minimum" approach - or at least an 80% kind of thing.

I think you really hit the nail on the head there.

I wouldn't stress 100% compatibility with every conceivable thing a keymap author can do. Instead, imho, what's important is:

1) Detecting incompatible syntax and showing an error, as clear as possible.
2) General usability of the whole thing. Making it as easy as possible for people to:
  a) make keymaps
  b) share their keymaps for troubleshooting (why doesn't this work for me?)
  c) share their results with the community

I know much of this is beyond the scope, just sharing some of my thoughts on what's important and what isn't. :)

**Comment by sboesebeck** [comment]
Hi Erez,

1) that’s a tough thing, as ELG parses the C-files directly, it has to ignore most of the file. That will cause problems of course...
2) I agree fully - not sure how to deal with that though. The creation of keymaps is ok for my taste, not exceptionally good, but I think one can manage. How I would solve the sharing thing, I just do not know. 

I actually think, that 2) is a bit out of scope for me being alone on this project. IF there were a lot of people using it and requesting such features, I could come up with some things (like using github for that). Right now it seems we're more or less the only ones playing with ELG...

**Comment by ezuk** [comment]
That makes total sense, on both counts. I suspect the project may see more traction soon, let's see what people ask and say. :)

**Comment by ezuk** [comment]
okay, back with some testing feedback!

1) Boot process is way, way better. I don't know if we can show an overlay that says "wait a moment" while it's doing all the stuff in the background, but anyway, it just worked seamlessly for me. Very cool to see it pull and compile and all that good stuff.

2) I loaded the default keymap and got a warning for unsupported syntax specifying the call. Perhaps we should make it clear that this is better for creating fresh keymaps, rather than modifying existing ones?

3) It compiled the default firmware! Then, ELG asked, "should I move the firmware file to the..." - but the dialog cut out there so I don't know where it moved it to. I hit okay, then went to the Teensy loader and File -> open, rummaged around a bit but couldn't find what I was looking for. So that's where my testing stopped for now, just don't know where my compiled keymap is.

4) Once we have the thing polished up, let's write a doc and make a video for it. And by "let's" I mean I will be happy to do it, with your feedback/approval on the contents.

**Comment by sboesebeck** [comment]
Hi Erez...

thanks a lot for the feedback...

let me answer on those a bit:

1) Not sure how to create a "please Wait" dialog in Linux itself. But there should be a way...
2) actually I wanted to use the ELG for reading in all the existing keymaps as well. There was a bug with the MEH_T and ALL_T checking, fixed that. But there is no Support for the FN-Keys. When I started this project, I don't think they were used. Do they have any advantage over doing the same with a macro? As I understand it, you use FN1 to momentarily toggle to the SYMB layer? Why not use the MO()-Call, which is supported by ELG...
3) the ELG compiles the firmware in qmk_firmware/keyboard/ergodox_ez/ directory. There, if all works, you end up with an ergodox-ez.hex file. If you answer the question about copying the file with "YES" it will copy the hex file in the keymap directory (e.g. keymaps/default_osx/default_osx.hex)
4) Sounds like a really great Idea - I'm the developer, I know how to do things and tend to describe them not in a proper way... thanks a lot!

The default_osx layout does have an inconsistency in it, causing the parsing to fail. The MDIA layer is not prepared, as it should... Usually you should have something like

[MDIA] = KEYMAP(....

but it is just KEYMAP(, no layer name given. This causes the parsing to fail unfortunately. If you just change that, it will work...  (I just did, see my pull request...). Also, the TG()-Call does not use the Layer name but the number instead. This makes parsing a bit hard and it might acutally fail. Changed that as well in the layout...

I would also change the FN-key usage to an MO-Call, but I'm not sure if there is a difference actually.

Cheers,

Stephan

**Comment by sboesebeck** [comment]
I'm just uploading a new Version of the VM, that has some minor improvements:

the Teensy loader is set to the right document (default compilation result)
the startup script was added to the ELG-Repository, so I can do changes without having to create a new VM-Image
added a banner output "please wait compiling sources" to give the waiting some sense...

As the qmk_sources are read from my clone of the repository, the default_osx layout is already "fixed" (except for the FN key)... go, check it out..

**Comment by sboesebeck** [comment]
Did you try the latest versions? there are a lot of fixes in there, making it more useable. Latest fix was including dynamic font sizes. Good for the generated png files and for the UI... 😉

**Comment by ezuk** [comment]
I am actually sending customers here right now :) I am at that point of the trial process -- want to see how it works for people in real life scenarios. Very excited about the progress!
